File: apps/user-desktop/src-tauri/src/runtime/bots.rs
/*
 * SPDX-License-Identifier: AGPL-3.0-only
 * Copyright (C) 2026 baibai and Botting contributors
 *
 * Botting is free software: you can redistribute it and/or modify it under
 * the GNU Affero General Public License version 3, as published by the
 * Free Software Foundation. This program comes WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the LICENSE file for the complete terms.
 * Copyleft: covered modifications must retain these license obligations.
 * https://www.gnu.org/licenses/agpl-3.0.html
 */

//! 啟動前更新即將到期的登入工作階段，再將必要資料交給 Bot runtime。

use super::UserRuntime;
use crate::{
    app_state::CommandResult,
    auth::{access_token, cookie, microsoft, ACCESS_TOKEN_EXPIRED, ACCESS_TOKEN_INVALID},
    bot_runtime::BotConfig,
    input_limits::{validate_length, MAX_BOT_ID_BYTES},
};
use chrono::{Duration, Utc};
use local_store::{AccountRecord, AuthKind};
use uuid::Uuid;

impl UserRuntime {
    /// 讀取帳號、更新到期工作階段後建立 Bot 連線。
    /// @param account_id 已保存的本機帳號 ID。
    /// @return 啟動請求結果；實際連線進度由 Bot 事件通知。
    pub(crate) async fn start_bot(&self, account_id: String) -> CommandResult<()> {
        validate_length(&account_id, "bot id", MAX_BOT_ID_BYTES)
            .map_err(|error| error.to_string())?;
        let mut account = self
            .with_store(move |store| {
                store
                    .account(&account_id)
                    .map_err(|error| error.to_string())?
                    .ok_or_else(|| "account not found".to_owned())
            })
            .await?;
        self.prepare_session(&mut account).await?;

        let profile_id = account
            .profile_id
            .as_deref()
            .ok_or_else(|| "Minecraft profile is missing".to_owned())?;
        let uuid = Uuid::parse_str(profile_id)
            .map_err(|_| "Minecraft profile UUID is invalid".to_owned())?;
        let access_token = active_access_token(&account)?;
        let config = BotConfig {
            bot_id: account.id,
            username: account.username,
            uuid,
            access_token,
            server_address: account.server_address,
        };
        self.bots
            .start(config)
            .await
            .map_err(|error| format!("start bot: {error:#}"))
    }

    /// 停止指定帳號的 Bot 工作階段。
    /// @param id 本機帳號 ID。
    /// @return 停止結果；未執行的 Bot 視為已停止。
    pub(crate) async fn stop_bot(&self, id: String) -> CommandResult<()> {
        validate_length(&id, "bot id", MAX_BOT_ID_BYTES).map_err(|error| error.to_string())?;
        self.bots.stop(&id).await.map_err(|error| error.to_string())
    }

    /// 依登入種類更新工作階段，並同步記憶體內帳號。
    /// @param account 即將用於連線的帳號，更新成功後原地修改。
    /// @return 工作階段可用或登入／儲存錯誤。
    async fn prepare_session(&self, account: &mut AccountRecord) -> CommandResult<()> {
        match account.auth_kind {
            AuthKind::Microsoft => self.prepare_microsoft(account).await,
            AuthKind::AccessToken => self.prepare_access_token(account).await,
            AuthKind::Cookie => self.prepare_cookie(account).await,
        }
    }

    async fn prepare_microsoft(&self, account: &mut AccountRecord) -> CommandResult<()> {
        if session_is_current(account) {
            return Ok(());
        }
        let session = microsoft::refresh(
            &self.http_client,
            account.credential.as_deref().unwrap_or_default(),
        )
        .await
        .map_err(|_| ACCESS_TOKEN_INVALID.to_owned())?;
        self.save_microsoft_session(account, session).await
    }

    async fn prepare_access_token(&self, account: &mut AccountRecord) -> CommandResult<()> {
        if account.session_token.is_some() {
            if session_is_current(account) {
                return Ok(());
            }
            let session = microsoft::refresh_legacy(
                &self.http_client,
                account.credential.as_deref().unwrap_or_default(),
            )
            .await
            .map_err(|_| ACCESS_TOKEN_INVALID.to_owned())?;
            return self.save_microsoft_session(account, session).await;
        }

        let expires_at = account
            .session_expires_at
            .or_else(|| account.credential.as_deref().and_then(access_token::expiry))
            .ok_or_else(|| ACCESS_TOKEN_INVALID.to_owned())?;
        if expires_at <= Utc::now() {
            return Err(ACCESS_TOKEN_EXPIRED.to_owned());
        }
        Ok(())
    }

    async fn prepare_cookie(&self, account: &mut AccountRecord) -> CommandResult<()> {
        if session_is_current(account) {
            return Ok(());
        }

        let session = cookie::exchange(account.credential.as_deref().unwrap_or_default())
            .await
            .map_err(|error| error.to_string())?;
        let account_id = account.id.clone();
        let username = session.username.clone();
        let uuid = session.uuid.clone();
        let access_token = session.access_token.clone();
        let expires_at = session.expires_at;
        self.with_store(move |store| {
            store
                .update_auth_session(
                    &account_id,
                    &username,
                    &uuid,
                    None,
                    &access_token,
                    Some(expires_at),
                )
                .map_err(|error| error.to_string())
        })
        .await?;
        account.username = session.username;
        account.profile_id = Some(session.uuid);
        account.session_token = Some(session.access_token);
        account.session_expires_at = Some(session.expires_at);
        Ok(())
    }

    /// 先保存完整登入結果，再更新呼叫端帳號快照。
    /// @param account 待同步的帳號。
    /// @param session 已驗證的新 Microsoft 工作階段。
    /// @return 儲存結果；寫入失敗不改動呼叫端快照。
    async fn save_microsoft_session(
        &self,
        account: &mut AccountRecord,
        session: microsoft::MicrosoftSession,
    ) -> CommandResult<()> {
        let account_id = account.id.clone();
        let username = session.username.clone();
        let uuid = session.uuid.clone();
        let refresh_token = session.refresh_token.clone();
        let access_token = session.access_token.clone();
        let expires_at = session.expires_at;
        self.with_store(move |store| {
            store
                .update_auth_session(
                    &account_id,
                    &username,
                    &uuid,
                    Some(&refresh_token),
                    &access_token,
                    Some(expires_at),
                )
                .map_err(|error| error.to_string())
        })
        .await?;
        account.username = session.username;
        account.profile_id = Some(session.uuid);
        account.credential = Some(session.refresh_token);
        account.session_token = Some(session.access_token);
        account.session_expires_at = Some(session.expires_at);
        Ok(())
    }
}

/// 優先使用工作階段權杖，無工作階段時才使用帳號憑據。
/// @param account 已完成登入準備的帳號。
/// @return 非空 access token；缺失時回傳固定錯誤代碼。
fn active_access_token(account: &AccountRecord) -> CommandResult<String> {
    account
        .session_token
        .clone()
        .or_else(|| account.credential.clone())
        .filter(|token| !token.trim().is_empty())
        .ok_or_else(|| ACCESS_TOKEN_INVALID.to_owned())
}

// 預留一分鐘登入緩衝，避免連線建立到一半才發現權杖到期。
/// 判斷現有權杖是否仍有至少一分鐘連線緩衝。
/// @param account 含工作階段權杖及期限的帳號。
/// @return 權杖存在且期限足夠時為 true。
fn session_is_current(account: &AccountRecord) -> bool {
    account.session_token.is_some()
        && account
            .session_expires_at
            .is_some_and(|expires_at| expires_at > Utc::now() + Duration::minutes(1))
}

File: apps/user-desktop/src/lib/features/bots.ts
/*
 * SPDX-License-Identifier: AGPL-3.0-only
 * Copyright (C) 2026 baibai and Botting contributors
 *
 * Botting is free software: you can redistribute it and/or modify it under
 * the GNU Affero General Public License version 3, as published by the
 * Free Software Foundation. This program comes WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the LICENSE file for the complete terms.
 * Copyleft: covered modifications must retain these license obligations.
 * https://www.gnu.org/licenses/agpl-3.0.html
 */

/** 處理 Bot 批次啟停、帳號操作與事件還原；每次更新前取得最新狀態。 */
import type { UserApi } from "../adapters/api";
import type {
  Account,
  BotPhase,
  MatchmakingSnapshot,
  RuntimeMode,
  SessionLogEntry,
  BotEvent,
} from "../models/types";

const knownBotErrorCodes = new Set([
  "banned",
  "connection_error",
  "kicked",
  "session_panic",
  "start_failed",
]);
const silentStatusPhases: ReadonlySet<BotPhase> = new Set(["starting", "connecting", "stopping"]);

export interface BotFeatureState {
  accounts: Account[];
  selectedBotIds: Set<string>;
  phases: Record<string, BotPhase>;
  activeMode: RuntimeMode;
  stoppingBotIds: Set<string>;
  matchmaking: MatchmakingSnapshot;
  matchmakingBusy: boolean;
  overlayVisible: boolean;
  selectedSessionBotId: string | null;
  deviceCode: { code: string; url: string } | null;
  deviceLinkError: string;
}

export interface BotFeatureDependencies {
  api: Pick<
    UserApi,
    | "startBot"
    | "stopBot"
    | "deleteAccounts"
    | "updateServer"
    | "stopMatchmaking"
    | "hideMatchmakingOverlay"
  >;
  getState: () => BotFeatureState;
  patchState: (patch: Partial<BotFeatureState>) => void;
  phaseLabel: (phase: BotPhase) => string;
  addLog: (message: string, botId?: string | null, level?: SessionLogEntry["level"]) => void;
  showToast: (message: string, kind: "success" | "error") => void;
}

/**
 * 建立帳號操作與 Bot 事件處理器。
 * @param deps API、即時狀態存取與通知回呼。
 * @return 批次啟停、帳號修改及事件還原介面。
 */
export function createBotFeature(deps: BotFeatureDependencies) {
  const recentErrorMessages = new Map<string, string>();

  /**
   * 平行啟動所選且離線的 Bot，個別失敗不取消整批。
   * @return 整批請求結束的 Promise；失敗會記錄至日誌。
   */
  async function startSelected(): Promise<void> {
    const state = deps.getState();
    const targets = state.accounts.filter((account) => {
      if (!state.selectedBotIds.has(account.id)) return false;
      const phase = state.phases[account.id] ?? "offline";
      return phase === "offline" || phase === "error";
    });
    if (!targets.length) return;

    const targetIds = new Set(targets.map((account) => account.id));
    deps.patchState({
      stoppingBotIds: new Set([...state.stoppingBotIds].filter((id) => !targetIds.has(id))),
    });
    deps.addLog(`starting ${targets.length} selected bots`);
    // 個別 Bot 啟動失敗不應取消同批其他 Bot 的啟動。
    const results = await Promise.allSettled(
      targets.map(async (account) => {
        const current = deps.getState();
        deps.patchState({
          phases: { ...current.phases, [account.id]: "starting" },
        });
        deps.addLog(`connecting to ${account.server_address}`, account.id);
        try {
          await deps.api.startBot(account.id);
        } catch (error) {
          const failed = deps.getState();
          deps.patchState({ phases: { ...failed.phases, [account.id]: "error" } });
          deps.addLog(error instanceof Error ? error.message : String(error), account.id, "error");
          throw error;
        }
      }),
    );
    const failed = results.filter((result) => result.status === "rejected").length;
    if (failed) deps.addLog(`${targets.length - failed}/${targets.length} selected bots started`);
  }

  /**
   * 停止目標 Bot，指令失敗時恢復尚未被新事件更新的狀態。
   * @param targets 本次要停止的帳號快照。
   * @return 整批停止流程的 Promise。
   */
  async function stopAccounts(targets: Account[]): Promise<void> {
    if (!targets.length) return;
    const state = deps.getState();
    deps.patchState({
      stoppingBotIds: new Set([...state.stoppingBotIds, ...targets.map((account) => account.id)]),
      phases: {
        ...state.phases,
        ...Object.fromEntries(targets.map((account) => [account.id, "stopping" as BotPhase])),
      },
    });
    const results = await Promise.allSettled(
      targets.map((account) => deps.api.stopBot(account.id)),
    );
    const current = deps.getState();
    const phases = { ...current.phases };
    const stoppingBotIds = new Set(current.stoppingBotIds);
    // 命令失敗時恢復原狀態；若事件已帶來更新，不可用舊快照蓋過它。
    results.forEach((result, index) => {
      const id = targets[index].id;
      stoppingBotIds.delete(id);
      if (phases[id] === "stopping") {
        phases[id] = result.status === "fulfilled" ? "offline" : (state.phases[id] ?? "offline");
      }
    });
    deps.patchState({ phases, stoppingBotIds });
    const failed = results.find(
      (result): result is PromiseRejectedResult => result.status === "rejected",
    );
    if (failed) deps.showToast(String(failed.reason), "error");
  }

  async function stopSelected(): Promise<void> {
    const state = deps.getState();
    await stopAccounts(
      state.accounts.filter((account) => {
        if (!state.selectedBotIds.has(account.id)) return false;
        const phase = state.phases[account.id] ?? "offline";
        return (
          phase === "online" ||
          phase === "starting" ||
          phase === "authenticating" ||
          phase === "connecting"
        );
      }),
    );
  }

  /**
   * 停止配對、隱藏浮動視窗並停止非離線 Bot。
   * @return 整個停止流程的 Promise。
   */
  async function stopSession(): Promise<void> {
    const state = deps.getState();
    const matchmaking = await deps.api.stopMatchmaking();
    deps.patchState({ matchmaking });
    await deps.api.hideMatchmakingOverlay();
    await stopAccounts(
      state.accounts.filter((account) => (state.phases[account.id] ?? "offline") !== "offline"),
    );
  }

  /**
   * 儲存非空且已變更的地址，再更新最新帳號清單。
   * @param account 被編輯的帳號。
   * @param serverAddress 輸入的新地址。
   * @return 保存及狀態同步的 Promise。
   */
  async function saveServerFor(account: Account, serverAddress: string): Promise<void> {
    const nextAddress = serverAddress.trim();
    if (!nextAddress || nextAddress === account.server_address) return;
    await deps.api.updateServer(account.id, nextAddress);
    const state = deps.getState();
    deps.patchState({
      accounts: state.accounts.map((item) =>
        item.id === account.id ? { ...item, server_address: nextAddress } : item,
      ),
    });
  }

  /**
   * 刪除後同步清理選取、階段與錯誤快取。
   * @param ids 待刪除的帳號 ID。
   * @return 後端刪除與畫面同步的 Promise。
   */
  async function deleteAccounts(ids: string[]): Promise<void> {
    if (!ids.length) return;
    await deps.api.deleteAccounts(ids);
    const state = deps.getState();
    const removed = new Set(ids);
    for (const id of removed) recentErrorMessages.delete(id);
    deps.patchState({
      accounts: state.accounts.filter((account) => !removed.has(account.id)),
      selectedBotIds: new Set([...state.selectedBotIds].filter((id) => !removed.has(id))),
      stoppingBotIds: new Set([...state.stoppingBotIds].filter((id) => !removed.has(id))),
      phases: Object.fromEntries(Object.entries(state.phases).filter(([id]) => !removed.has(id))),
      selectedSessionBotId:
        state.selectedSessionBotId && removed.has(state.selectedSessionBotId)
          ? null
          : state.selectedSessionBotId,
    });
  }

  /**
   * 將後端事件還原成 UI 狀態與去重後的日誌。
   * @param event 一筆 Bot 事件。
   * @return 無回傳值；透過 patchState 發布更新。
   */
  function handleBotEvent(event: BotEvent): void {
    const state = deps.getState();
    if (event.type === "status") {
      // 停止指令完成前，忽略同一 Bot 尚未排出的舊狀態事件。
      if (state.stoppingBotIds.has(event.bot_id) && event.phase !== "offline") return;
      if (
        event.phase === "starting" ||
        event.phase === "authenticating" ||
        event.phase === "connecting" ||
        event.phase === "online"
      ) {
        recentErrorMessages.delete(event.bot_id);
      }
      const previousPhase = state.phases[event.bot_id];
      const repeatsVisiblePhase =
        previousPhase !== undefined &&
        previousPhase !== event.phase &&
        !event.message &&
        deps.phaseLabel(previousPhase) === deps.phaseLabel(event.phase);
      const repeatsError =
        event.phase === "offline" &&
        event.message !== null &&
        event.message !== undefined &&
        recentErrorMessages.get(event.bot_id) === event.message;
      const stoppingBotIds =
        event.phase === "offline"
          ? new Set([...state.stoppingBotIds].filter((id) => id !== event.bot_id))
          : state.stoppingBotIds;
      deps.patchState({
        stoppingBotIds,
        phases: { ...state.phases, [event.bot_id]: event.phase },
      });
      if (!silentStatusPhases.has(event.phase) && !repeatsVisiblePhase && !repeatsError) {
        const message =
          event.message === "Stopped from Botting"
            ? "Stopped"
            : event.message || deps.phaseLabel(event.phase);
        deps.addLog(message, event.bot_id, event.phase === "error" ? "error" : "info");
      }
    } else if (event.type === "runtime_mode") {
      deps.patchState({ activeMode: event.mode });
    } else if (event.type === "profile") {
      deps.patchState({
        accounts: state.accounts.map((account) =>
          account.id === event.bot_id
            ? {
                ...account,
                username: event.username,
                credential_checked_at: new Date().toISOString(),
              }
            : account,
        ),
      });
    } else if (event.type === "device_code") {
      deps.patchState({
        deviceCode: { code: event.user_code, url: event.verification_uri },
        deviceLinkError: "",
      });
    } else if (event.type === "microsoft_auth_result") {
      deps.patchState({ deviceCode: null });
    } else if (event.type === "afk_state") {
      deps.addLog(event.active ? "AFK started" : "AFK stopped", event.bot_id);
    } else if (event.type === "match_attempt_failed") {
      if (!isCommandSpamFailure(event.code, event.message)) {
        deps.addLog(`Matchmaking failed: ${event.message}`, event.bot_id, "error");
      }
    } else if (event.type === "chat_message") {
      deps.addLog(`[chat] ${event.message}`, event.bot_id);
    } else if (event.type === "matchmaking_debug") {
      deps.addLog(`[match debug] ${event.message}`, event.bot_id);
    } else if (event.type === "error") {
      if (event.bot_id && state.stoppingBotIds.has(event.bot_id)) return;
      deps.patchState({
        phases: event.bot_id ? { ...state.phases, [event.bot_id]: "error" } : state.phases,
      });
      const message = botErrorMessage(event.code, event.message);
      if (event.bot_id) recentErrorMessages.set(event.bot_id, message);
      deps.addLog(message, event.bot_id ?? null, "error");
    }
  }

  return {
    deleteAccounts,
    handleBotEvent,
    saveServerFor,
    startSelected,
    stopAccounts,
    stopSelected,
    stopSession,
  };
}

function botErrorMessage(code: string, message: string): string {
  return knownBotErrorCodes.has(code) ? message : `${code}: ${message}`;
}

function isCommandSpamFailure(code: string, message: string): boolean {
  return (
    code === "command_spam" ||
    /please don't spam the command|sending commands too fast/i.test(message)
  );
}
